test/BSCValidatorSet.js

```diff
@@ -600,6 +600,133 @@ contract('BSCValidatorSet', (accounts) => {
+      newValidators.push(web3.eth.accounts.create().address)
+    }
+    let packageBytes = validatorUpdateRlpEncode(newValidators,
+        newValidators, newValidators);
+    await validatorSetInstance.handleSynPackage(STAKE_CHANNEL_ID, packageBytes, {from: relayerAccount});
+
+    // set numOfCabinets to 21
+    let govChannelSeq = await crossChain.channelReceiveSequenceMap(GOV_CHANNEL_ID);
+    let govValue = "0x0000000000000000000000000000000000000000000000000000000000000015";// 21;
+    let govPackageBytes = serializeGovPack("numOfCabinets", govValue, validatorSetInstance.address);
+    await crossChain.handlePackage(Buffer.concat([buildSyncPackagePrefix(2e16), (govPackageBytes)]), proof, merkleHeight, govChannelSeq, GOV_CHANNEL_ID, {from: relayer});
+
+    let except = await validatorSetInstance.numOfCabinets.call();
+    assert.equal(web3.utils.toBN(except).eq(web3.utils.toBN(21)), true, "wrong numOfCabinets");
+
+    // without candidate validators
+    let maxNumOfWorkingCandidates = 2;
+    let numOfCabinets = 21;
+    let validators = await validatorSetInstance.getValidators.call();
+    let miningValidators = await validatorSetInstance.getMiningValidators.call();
+    assert.deepEqual(validators.slice(0,numOfCabinets), miningValidators, "wrong validators");
+
+    // set maxNumOfCandidates to 20
+    govChannelSeq = await crossChain.channelReceiveSequenceMap(GOV_CHANNEL_ID)
+    govValue = "0x0000000000000000000000000000000000000000000000000000000000000014";// 20;
+    govPackageBytes = serializeGovPack("maxNumOfCandidates", govValue, validatorSetInstance.address);
+    await crossChain.handlePackage(Buffer.concat([buildSyncPackagePrefix(2e16), (govPackageBytes)]), proof, merkleHeight, govChannelSeq, GOV_CHANNEL_ID, {from: relayer});
+
+    except = await validatorSetInstance.maxNumOfCandidates.call();
+    assert.equal(web3.utils.toBN(except).eq(web3.utils.toBN(20)), true, "wrong maxNumOfCandidates");
+
+    // set maxNumOfWorkingCandidates to 2
+    govChannelSeq = await crossChain.channelReceiveSequenceMap(GOV_CHANNEL_ID)
+    govValue = "0x0000000000000000000000000000000000000000000000000000000000000002";// 2;
+    govPackageBytes = serializeGovPack("maxNumOfWorkingCandidates", govValue, validatorSetInstance.address);
+    await crossChain.handlePackage(Buffer.concat([buildSyncPackagePrefix(2e16), (govPackageBytes)]), proof, merkleHeight, govChannelSeq, GOV_CHANNEL_ID, {from: relayer});
+
+    except = await validatorSetInstance.maxNumOfWorkingCandidates.call();
+    assert.equal(web3.utils.toBN(except).eq(web3.utils.toBN(2)), true, "wrong maxNumOfWorkingCandidates");
+
+    if ((validators.length - numOfCabinets) < maxNumOfWorkingCandidates){
+      maxNumOfWorkingCandidates = validators.length - numOfCabinets;
+    } 
+    
+    miningValidators = await validatorSetInstance.getMiningValidators.call();
+    let exceptValues = validators.slice(0,numOfCabinets);
+    for (var j=0;j<numOfCabinets-maxNumOfWorkingCandidates;j++){
```

Comment:
What about the `miningValidators[numOfCabinets-maxNumOfWorkingCandidates: numOfCabinets]`, can you come up with any idea to ensure that candidates can be one of them?